storage/zynq7000-flash/flashdrv.c

```diff
@@ -0,0 +1,502 @@
+	qspi_stop();
+
+	return res;
+}
+
+
+static ssize_t flashdrv_pageProgram(addr_t offs, const void *buff, size_t len)
+{
+	ssize_t res;
+	time_t timeout;
+	const flash_cfi_t *cfi = &fdrv_common.info.cfi;
+	const flash_cmd_t *cmd = &fdrv_common.info.cmds[flash_cmd_pp];
+
+	if (len == 0)
+		return 0;
+
+	if (buff == NULL || len % CFI_SIZE_PAGE(cfi->pageSize))
+		return -EINVAL;
+
+	if ((res = flashdrv_welSet(flash_cmd_wren)) < 0)
+		return res;
+
+	memset(fdrv_common.cmdTx, 0, cmd->size);
+	fdrv_common.cmdTx[0] = cmd->opCode;
+	fdrv_common.cmdTx[1] = (offs >> 16) & 0xff;
+	fdrv_common.cmdTx[2] = (offs >> 8) & 0xff;
+	fdrv_common.cmdTx[3] = offs & 0xff;
+
+	qspi_start();
+	if ((res = qspi_transfer(fdrv_common.cmdTx, fdrv_common.cmdRx, cmd->size, TIMEOUT_CMD_MS)) < 0) {
+		qspi_stop();
+		return res;
+	}
+
+	/* Transmit data to write */
+	if ((res = qspi_transfer(buff, NULL, len, TIMEOUT_CMD_MS)) < 0) {
+		qspi_stop();
+		return res;
+	}
+	qspi_stop();
+
+	len = res;
+
+	timeout = CFI_TIMEOUT_MAX_PROGRAM(cfi->timeoutTypical.pageWrite, cfi->timeoutMax.pageWrite) + TIMEOUT_CMD_MS;
+	if ((res = flashdrv_wipCheck(timeout)) < 0)
+		return res;
+
+	return len;
+}
+
+
+/* Device interface */
+
+void flashdrv_flashInfo(flash_info_t *info)
+{
+	memcpy(info, &fdrv_common.info, sizeof(flash_info_t));
+}
+
+
+int flashdrv_sync(void)
+{
+	uint8_t *src;
+	addr_t dst;
+	ssize_t res;
+	unsigned int i, pgNb;
+
+	size_t pageSz, sectSz, regStart;
+	const flash_cfi_t *cfi = &fdrv_common.info.cfi;
+
+	if (fdrv_common.buffPos == 0)
+		return EOK;
+
+	regStart = flashdrv_regStart(fdrv_common.regID);
+	sectSz = CFI_SIZE_SECTION(cfi->regs[fdrv_common.regID].size);
+	pageSz = CFI_SIZE_PAGE(cfi->pageSize);
+	pgNb = sectSz / pageSz;
+
+	for (i = 0; i < pgNb; ++i) {
+		dst = regStart + fdrv_common.sectID * sectSz + i * pageSz;
+		src = fdrv_common.buff + i * pageSz;
+
+		if ((res = flashdrv_pageProgram(dst, src, pageSz)) < 0)
+			return res;
+	}
+
+	fdrv_common.regID = -1;
+	fdrv_common.sectID = -1;
+	fdrv_common.buffPos = 0;
+
+	return EOK;
+}
+
+
+int flashdrv_sectorErase(addr_t offs)
+{
+	int i;
+	ssize_t res;
+	size_t sectorSz;
+	time_t timeout;
+	const flash_cmd_t *cmd;
+	const flash_cfi_t *cfi = &fdrv_common.info.cfi;
+
+	if (offs >= CFI_SIZE_FLASH(cfi->chipSize))
+		return -EINVAL;
+
+	if ((i = flashdrv_regionFind(offs)) < 0)
+		return i;
+
+	/* Check offset alligment */
+	sectorSz = CFI_SIZE_SECTION(cfi->regs[i].size);
+	if (offs & (sectorSz - 1))
+		return -EINVAL;
+
+	switch (sectorSz) {
+		case 0x1000:
+			cmd = &fdrv_common.info.cmds[flash_cmd_p4e];
+			break;
+		case 0x10000:
+			cmd = &fdrv_common.info.cmds[flash_cmd_p64e];
+			break;
+		default:
+			return -EINVAL;
+	}
+
+	if ((res = flashdrv_welSet(flash_cmd_wren)) < 0)
+		return res;
+
+	memset(fdrv_common.cmdTx, 0, cmd->size);
+	fdrv_common.cmdTx[0] = cmd->opCode;
+	fdrv_common.cmdTx[1] = (offs >> 16) & 0xff;
+	fdrv_common.cmdTx[2] = (offs >> 8) & 0xff;
+	fdrv_common.cmdTx[3] = offs & 0xff;
```

Comment:
I see this construction often. Maybe static inline this?

Reply:
Done